File: backend/src/domain/repository/IResourceRepository.ts
import { LearningResource } from '../entities/LearningResource';
import { Job } from '../entities/Job';

/**
 * Repository interface for learning resources and job posting operations
 */
export interface IResourceRepository {
  // Learning Resource Management
  createLearningResource(
    resourceData: Partial<LearningResource>,
  ): Promise<LearningResource>;
  updateLearningResource(
    resourceId: number,
    resourceData: Partial<LearningResource>,
  ): Promise<LearningResource>;
  deleteLearningResource(resourceId: number): Promise<boolean>;
  findLearningResourceById(
    resourceId: number,
  ): Promise<LearningResource | null>;
  findLearningResourcesByOrganization(
    organizationId: number,
    page: number,
    limit: number,
    filters?: { type?: string },
  ): Promise<{ resources: LearningResource[]; total: number }>;
  findLearningResourcesByUploader(
    uploaderMemberId: number,
    page: number,
    limit: number,
  ): Promise<{ resources: LearningResource[]; total: number }>;
  searchLearningResources(
    organizationId: number,
    keyword: string,
    page: number,
    limit: number,
  ): Promise<{ resources: LearningResource[]; total: number }>;

  // Job Management
  createJob(jobData: Partial<Job>): Promise<Job>;
  updateJob(jobId: number, jobData: Partial<Job>): Promise<Job>;
  deleteJob(jobId: number): Promise<boolean>;
  findJobById(jobId: number): Promise<Job | null>;
  findJobsByOrganization(
    organizationId: number,
    page: number,
    limit: number,
    filters?: { type?: string; isActive?: boolean; isReferral?: boolean },
  ): Promise<{ jobs: Job[]; total: number }>;
  findJobsByPoster(
    posterMemberId: number,
    page: number,
    limit: number,
  ): Promise<{ jobs: Job[]; total: number }>;
  searchJobs(
    organizationId: number,
    keyword: string,
    page: number,
    limit: number,
  ): Promise<{ jobs: Job[]; total: number }>;
  findActiveJobs(
    organizationId: number,
    page: number,
    limit: number,
  ): Promise<{ jobs: Job[]; total: number }>;
  activateJob(jobId: number): Promise<Job>;
  deactivateJob(jobId: number): Promise<Job>;
  findJobsByDeadline(
    organizationId: number,
    beforeDate: Date,
    page: number,
    limit: number,
  ): Promise<{ jobs: Job[]; total: number }>;

  // Statistics
  getResourceStatistics(organizationId: number): Promise<{
    totalResources: number;
    resourcesByType: { type: string; count: number }[];
  }>;

  getJobStatistics(organizationId: number): Promise<{
    totalJobs: number;
    activeJobs: number;
    referralJobs: number;
    jobsByType: { type: string; count: number }[];
  }>;
}

File: backend/src/main/java/com/service/backend/domain/repository/IResourceRepository.java
package com.service.backend.domain.repository;

import com.service.backend.domain.entity.LearningResource;
import com.service.backend.domain.entity.Job;
import reactor.core.publisher.Mono;

import java.time.LocalDate;
import java.util.Map;

/**
 * Repository interface for learning resources and job posting operations
 */
public interface IResourceRepository {

    // Learning Resource Management
    Mono<LearningResource> createLearningResource(LearningResource resourceData);
    Mono<LearningResource> updateLearningResource(Long resourceId, LearningResource resourceData);
    Mono<Boolean> deleteLearningResource(Long resourceId);
    Mono<LearningResource> findLearningResourceById(Long resourceId);
    Mono<Map<String, Object>> findLearningResourcesByOrganization(Long organizationId, int page, int limit, Map<String, Object> filters);
    Mono<Map<String, Object>> findLearningResourcesByUploader(Long uploaderMemberId, int page, int limit);
    Mono<Map<String, Object>> searchLearningResources(Long organizationId, String keyword, int page, int limit);

    // Job Management
    Mono<Job> createJob(Job jobData);
    Mono<Job> updateJob(Long jobId, Job jobData);
    Mono<Boolean> deleteJob(Long jobId);
    Mono<Job> findJobById(Long jobId);
    Mono<Map<String, Object>> findJobsByOrganization(Long organizationId, int page, int limit, Map<String, Object> filters);
    Mono<Map<String, Object>> findJobsByPoster(Long posterMemberId, int page, int limit);
    Mono<Map<String, Object>> searchJobs(Long organizationId, String keyword, int page, int limit);
    Mono<Map<String, Object>> findActiveJobs(Long organizationId, int page, int limit);
    Mono<Job> activateJob(Long jobId);
    Mono<Job> deactivateJob(Long jobId);
    Mono<Map<String, Object>> findJobsByDeadline(Long organizationId, LocalDate beforeDate, int page, int limit);

    // Statistics
    Mono<Map<String, Object>> getResourceStatistics(Long organizationId);
    Mono<Map<String, Object>> getJobStatistics(Long organizationId);
}
